### PR TITLE
Feature/update to py3/pipeline

### PR DESCRIPTION
**Pull request by WilliamHarvey97**
This PR adds a pipeline feature. It enables the user to create a pipeline wich is (of course) multithreaded. This pipeline automatically link queues between threads and can be shutted down anytime.

### CODE REVIEW

**Review by AlexandreMarcotte — APPROVED** [review]
Good code, nothing to add.

File: pymuse/inputstream/muse_osc_input_stream.py [hunk 1]
```diff
@@ -1,32 +1,38 @@
...
 from pymuse.signal import Signal, SignalData
 from pymuse.inputstream.constants import DEFAULT_UDP_PORT, LOCALHOST, SIGNAL_QUEUE_LENGTH
 from pymuse.inputstream.muse_constants import (
     MUSE_ACQUISITION_FREQUENCIES,
-    MUSE_BATT_ACQUISITION_FREQUENCY,
-    MUSE_EEG_ACQUISITION_FREQUENCY,
     MUSE_OSC_PATH,
 )
 
+from time import time
```
Comment (conorato) [inline pymuse/inputstream/muse_osc_input_stream.py]: unused import

File: pymuse/pipelinestages/pipeline_stage.py [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+    def _write_queues_out(self, data):
+        for queue_out in self._queues_out:
+            queue_out.put(deepcopy(data))
+
+    # Override this method to correctly shutdown your pipeline stage
+    def shutdown(self):
+        """Kill the current module."""
+        self._shutdown_hook()
+        self._shutdown_event.set()
+        self._shutdown_event.wait()
```
Comment (conorato) [inline pymuse/pipelinestages/pipeline_stage.py]: wait is useless in this case

File: pymuse/inputstream/test/test_input_stream.py [hunk 3]
```diff
@@ -24,16 +24,22 @@ def read_messages(self, messages: list, signal_name: str):
...
-        Thread(self.create_client(SOUGHT_DATA_LIST[1], TOUCHING_FOREHEAD_MESSAGES_LIST)).start()
+        Thread(self.create_client(
+            SOUGHT_DATA_LIST[0], EEG_MESSAGES_LIST)).start()
+        Thread(self.create_client(
+            SOUGHT_DATA_LIST[1], TOUCHING_FOREHEAD_MESSAGES_LIST)).start()
         self.read_messages(EEG_MESSAGES_LIST, SOUGHT_DATA_LIST[0])
-        self.read_messages(TOUCHING_FOREHEAD_MESSAGES_LIST, SOUGHT_DATA_LIST[1])
+        self.read_messages(TOUCHING_FOREHEAD_MESSAGES_LIST,
+                           SOUGHT_DATA_LIST[1])
         self.muse_input_stream.close()
```
Comment (conorato) [inline pymuse/inputstream/test/test_input_stream.py]: Unittest didnt pass cause 'close' method was switched to 'shutdown'

File: pymuse/pipeline.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+                for forked_branch in stage.forked_branches:
+                    self._start(forked_branch)
+            else:
+                stage.start()
+    
+    def get_output_queue(self, queue_index = 0) -> Queue:
+        """Return a ref to the queue given by queue_index"""
+        return self._output_queues[queue_index]
+
+    def read_output_queue(self, queue_index = 0):
```
Comment (conorato) [inline pymuse/pipeline.py]: missing return type of method

File: pymuse/pipeline.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    def __init__(self, input_signal: Signal, *stages):
+        """ E.g.: Pipeline(Signal(), PipelineStage(), PipelineFork([PipelineStage(), PipelineStage()], [PipelineStage()] )) """
+        self._output_queues = []
+        self._input_signal = input_signal
+        self._stages: list = list(stages)
+        self._link_stages(self._stages)
+        self._stages[0]._queue_in = self._input_signal.signal_queue
+
+    def _link_pipeline_fork (self, stages: list, index: int):
+            for fork in stages[index].forked_branches:
```
Comment (conorato) [inline pymuse/pipeline.py]: wrong indentation

File: pymuse/pipeline.py [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+    def _shutdown(self, stages: list):
+        for stage in stages:
+            if type(stage) == PipelineFork:
+                for forked_branch in stage.forked_branches:
+                    self._shutdown(forked_branch)
+            else:
+                stage.shutdown()
+
+    
+    def shutdown(self):
```
Comment (conorato) [inline pymuse/pipeline.py]: public methods should be at the top of the class, before defining private methods

File: pymuse/test/test_pipeline.py [hunk 7]
```diff
@@ -0,0 +1,51 @@
+from pymuse.pipelinestages.pipeline_stage import PipelineStage
```
Comment (conorato) [inline pymuse/test/test_pipeline.py:4]: general package import should be before package specific import

**Review by conorato — COMMENTED** [review]
We could add docstring summary to every class to ensure documentation throughout the code.

File: pymuse/inputstream/test/test_input_stream.py (hunk 3 above)
Comment (WilliamHarvey97) [reply pymuse/inputstream/test/test_input_stream.py]: TY

File: pymuse/pipeline.py (hunk 4 above)
Comment (WilliamHarvey97) [reply pymuse/pipeline.py]: Cannot say it's alway's gonna be a SignalData that is passed between PipelineStages

**Review by conorato — APPROVED** [review]
Good job William

**Pull request merged**